### PR TITLE
Improve shared key derivation and add more test vectors

### PR DESCRIPTION
**Pull request by ttaubert**

### CODE REVIEW

File: draft-bar-cfrg-spake2plus.md [hunk 1]
```diff
@@ -312,20 +312,27 @@ received.
...
 The protocol transcript TT, as defined in {{online-authentication}}, is unique and secret to
-the participants. Both parties use TT to derive shared symmetric secrets K_auth
-and K_enc and output K_enc as the shared secret from the protocol. The length of
-each key is equal to half of the digest output, e.g., |K_auth| = |K_enc| = 128
-bits for Hash() = SHA-256.
+the participants. Both parties use TT to derive the shared symmetric secret K_main from the
+protocol. The length of K_main is equal to the length of the digest output, e.g., 256 bits
+for Hash() = SHA-256. The confirmation keys K_confirmP and K_confirmV, as well as the shared
+encryption key K_shared are derived from K_main. The length of K_shared can be freely chosen
+by the application.
```
Comment (chris-wood) [inline draft-bar-cfrg-spake2plus.md:319]:
```suggestion
encryption key K_shared are derived from K_main.
```

Comment (chris-wood) [inline draft-bar-cfrg-spake2plus.md:319]:
```suggestion
encryption key K_shared are derived from K_main. The length of K_shared is equal to the
length of the KDF output, which is defined by the ciphersuite.
```

File: draft-bar-cfrg-spake2plus.md [hunk 2]
```diff
@@ -312,20 +312,27 @@ received.
...
+K_shared = KDF(nil, K_main, "SharedKey")
 ~~~
 
-K_auth is used to derive K_confirmP and K_confirmV. The length of each confirmation key
-is equal to half of the digest output, e.g., |K_confirmP| = |K_confirmV| = 128 bits for
-Hash() = SHA-256. Neither K_auth nor its derived confirmation keys are used for
-anything except key confirmation and MUST be discarded after the protocol execution.
+Neither K_main nor its derived confirmation keys are used for anything except key
+derivation and confirmation and MUST be discarded after the protocol execution.
+Besides K_shared, applications MAY derive additional keys from K_main as needed.
```
Comment (chris-wood) [inline draft-bar-cfrg-spake2plus.md:329]:
```suggestion
Applications MAY derive additional keys from K_shared as needed.
```

File: draft-bar-cfrg-spake2plus.md [hunk 3]
```diff
@@ -312,20 +312,27 @@ received.
...
+
+The length of each confirmation key is dependent on the MAC function of the chosen
+ciphersuite. For HKDF, the RECOMMENDED key length is equal to the output length of
```
Comment (chris-wood) [inline draft-bar-cfrg-spake2plus.md:332]:
```suggestion
ciphersuite. For HMAC, the RECOMMENDED key length is equal to the output length of
```

**Review by chris-wood — CHANGES_REQUESTED** [review]
Some editorial changes requested, but otherwise this is great. Nice work with the reference implementation!

**Review by chris-wood — APPROVED** [review]
LGTM pending one final nit!

File: draft-bar-cfrg-spake2plus.md [hunk 4]
```diff
@@ -312,20 +312,26 @@ received.
...
 
 The protocol transcript TT, as defined in {{online-authentication}}, is unique and secret to
-the participants. Both parties use TT to derive shared symmetric secrets K_auth
-and K_enc and output K_enc as the shared secret from the protocol. The length of
-each key is equal to half of the digest output, e.g., |K_auth| = |K_enc| = 128
-bits for Hash() = SHA-256.
+the participants. Both parties use TT to derive the shared symmetric secret K_main from the
+protocol. The length of K_main is equal to the length of the digest output, e.g., 256 bits
+for Hash() = SHA-256. The confirmation keys K_confirmP and K_confirmV, as well as the shared
+encryption key K_shared are derived from K_main.
```
[the remark this answers is not in the document]
Comment (ttaubert) [reply draft-bar-cfrg-spake2plus.md:318]: Ah, yes. Good spot!

**Pull request merged**